### PR TITLE
Data output improvements

### PR DESCRIPTION
**Pull request by adriaan**
This PR adds the following improvements to the data output of the notebook:

**There are now two ways to export CSVs:**

1. A single CSV containing all energy data for all devices, accompanied with a CSV containing the metadata for all devices (which can e.g. be used to map channel categories to the energy data).
2. One CSV per device, with the CSV headers containing the channel category name.

**Option to include date and time to the energy data**

In addition to the UNIX timestamp, a date and time field are added to the exported data (both JSON and CSV). The date and time are based on the timestamp and the device's timezone (or the fallback defined in the configuration if there is no timezone data available for the device)

### CODE REVIEW

File: wattwatchers_rest_api_v3_long_energy.ipynb [hunk 1]
```diff
@@ -636,6 +633,7 @@
         "        writer.writerow(row)\n",
         "\n",
         "# Method to save metadata for all devices to a CSV file on Google Drive\n",
+        "# The metedata can be used to map channel category labels for a device to its energy data, amongst other things.\n",
```
Comment (graceyoung-ww) [inline wattwatchers_rest_api_v3_long_energy.ipynb]: Typo in comment: "metadata"

File: wattwatchers_rest_api_v3_long_energy.ipynb [hunk 2]
```diff
@@ -381,7 +373,16 @@
         "    return results\n",
         "\n",
         "  # Helper method to generate a dictionary with an id, data and (optionally) device metadata\n",
-        "  def device_le_dict_for(device_id, data, metadata):\n",
+        "  def device_le_dict_for(device_id, timezone, data, metadata):\n",
+        "    if INCLUDE_DATE_AND_TIME:\n",
+        "      tz = pytz.timezone(timezone)\n",
+        "      def _include_date_and_time(interval: dict) -> dict:\n",
+        "        datetime = dt.datetime.fromtimestamp(interval['timestamp'], tz)\n",
+        "        interval['date'] = datetime.strftime('%m/%d/%Y')\n",
```
Comment (graceyoung-ww) [inline wattwatchers_rest_api_v3_long_energy.ipynb:381]: Assuming the MM/DD format is what Excel expects?

File: wattwatchers_rest_api_v3_long_energy.ipynb [hunk 3]
```diff
@@ -410,28 +411,28 @@
...
+      "metadata": {
+        "id": "iO0dXSWLASYh"
+      },
+      "outputs": [],
       "source": [
-        "import datetime as dt\n",
-        "import pytz\n",
-        "\n",
         "# Convenience method to convert a start date and end date string\n",
         "# into a end time stamp and interval that can be used as parameters for the API methods.\n",
```
Comment (graceyoung-ww) [inline wattwatchers_rest_api_v3_long_energy.ipynb]: `timestamp` one word?

File: wattwatchers_rest_api_v3_long_energy.ipynb [hunk 4]
```diff
@@ -410,28 +411,28 @@
...
         "# Start and end date parameters are to be formatted as '2021-06-29 17:08:00' (or according tot he date_format parameter passed in)\n",
```
Comment (graceyoung-ww) [inline wattwatchers_rest_api_v3_long_energy.ipynb]: Typo `to the`

Comment (graceyoung-ww) [reply wattwatchers_rest_api_v3_long_energy.ipynb]: I just had the thought, perhaps this should be the https://en.wikipedia.org/wiki/ISO_8601#:~:text=As%20of%20ISO%208601-1,minute%20between%2000%20and%2059. format, which is the same as you have but the space is replaced by `T`

File: wattwatchers_rest_api_v3_long_energy.ipynb [hunk 5]
```diff
@@ -504,43 +502,123 @@
...
+        "      if len(device['data']) > 0:\n",
+        "        return device\n",
+        "    return None\n",
+        "  \n",
+        "  # determine csv headers based on device data from device that has energy data\n",
+        "  device_with_data = _first_device_with_data(data)\n",
+        "  if device_with_data:\n",
+        "    headers = list(_flatten_arrays(device_with_data['data'][0]).keys())\n",
+        "  else:\n",
+        "    print('None of the devices have associated energy data. CSV export aborted')\n",
```
Comment (graceyoung-ww) [inline wattwatchers_rest_api_v3_long_energy.ipynb]: `CSV export aborted.` (missing period)

File: wattwatchers_rest_api_v3_long_energy.ipynb [hunk 6]
```diff
@@ -554,38 +632,77 @@
...
+        "  \n",
+        "  with open(f'/gdrive/My Drive/{file_name}.csv', 'w', newline='') as f:\n",
+        "    writer = csv.DictWriter(f, headers)\n",
+        "    writer.writeheader()\n",
+        "\n",
+        "    for device in data:\n",
+        "      row = _device_row_from_device_data(device)\n",
+        "      writer.writerow(row)\n",
+        "\n",
         "# Method to save data to file on Google Drive (either in CSV or JSON format)\n",
```
Comment (graceyoung-ww) [inline wattwatchers_rest_api_v3_long_energy.ipynb:658]: Nice that this supports JSON as well!

**Review by graceyoung-ww — APPROVED** [review]

File: wattwatchers_rest_api_v3_long_energy.ipynb (hunk 2 above)
Comment (adriaan) [reply wattwatchers_rest_api_v3_long_energy.ipynb:381]: Yeah, tested importing in Excel (after some Googling for supported date formats). This worked, so left it at that.

**Pull request merged**